rfc/system/5598-deterministic-errors-distributed-training.md

```diff
@@ -0,0 +1,93 @@
+
+For these distributed jobs, multiple Kubernetes pods are launched. Any of these worker pods can crash, causing all other worker pods in the distributed job to fail subsequently because one worker disappeared.
+
+Error propagation, in Flyte, happens by the pod entrypoint uploading a file called `error.pb` to blob storage which contains (among other things) the error message and the information whether the error is retriable.
+
+In a distributed training job, all worker pods currently try to create the same `error.pb` file in blob storage - leading to a race condition. It is not guaranteed that the root-cause error is the one being reported to the user and used to determine whether the task can be retried.
+
+## 2 Motivation
+
+* As a Flyte user trying to understand why a distributed training task failed, I currently cannot rely on the error reported in the Flyte Console (UI) being the root cause error.
+    * Instead, I have to search the logs of each worker pod. For distributed training jobs with dozens or even hundreds of worker pods, this can be tedious.
+    * (Current remedies include combining all worker pods in stackdriver logs using a wildcard in the pod name and then filtering by severity.)
+* As a Flyte user marking specific errors that can occur in distributed training jobs as retriable (using a `FlyteRecoverableException`), I want Flyte to deterministically determine the root cause error so that the retry behaviour does not suffer from a race condition.
+
+## 3 Proposed Implementation
+
+### Flytekit
+
+#### Preventing the race condition
+
+For distributed training tasks, the [pod entrypoint `pyflyte-execute`](https://github.com/flyteorg/flytekit/blob/master/flytekit/bin/entrypoint.py) must not upload a file called [`error.pb`](https://github.com/flyteorg/flytekit/blob/77d056ab9fda40ec6b2312a4d197b9107cdb70dc/flytekit/core/constants.py#L4) (which is the same for all worker pods) but instead choose a file name which differs for each worker pod. We propose to simply include the pod name in the `error-<pod name>.pb`. This prevents the race condition and has the added benefit that with this information displayed in the error message in the UI, it is easy to inspect the problematic pod.
+
+Open questions:
+* How does the pod entrypoint determine that a specific task plugin requires separate error files for each worker pod?
+    * One of the task base classes, e.g. `PythonFunctionTask` (at which level should we do this?) could have an attribute `is_distributed` which is set to `False`. Distributed training tasks which inherit from this base class would overwrite this attribute to true. The entrypoint would check this attribute to determine the correct naming of the error file.
+* Should we not configure this in the task classes in flytekit but inject this information e.g. via an env var in `flyteplugins` (backend)?
```

Comment:
I assume Flyte backend knows whether a task is being implemented by multiple pods or not. If this information is available, it is a better design IMO not to burden the task plugin developers with remembering to override is_distributed.

Reply:
Yes, the respective plugins in `flyteplugins` know and could inject an env var. We already set env vars such as `FLYTE_INTERNAL_EXECUTION_ID`, ... What do you think of adding `FLYTE_INTERNAL_ERROR_PROPAGATION=earliest`? (Not making it a binary flag leaves room for potential other future strategies.)

If we do it this way, backwards compatibility also becomes easier:
If the flytekit entrypoint reacts to the env var and if the to be added multi-error-file-output-reader checks for the default `error.pb` as a fallback, all cases (flytekit too old, backend too old, ...) should be covered.

Reply:
If we use the pod name in the error file, we also have to set `FLYTE_INTERNAL_POD_NAME` via the downward api.

Reply:
Agreed, this makes backward compatibility easier.